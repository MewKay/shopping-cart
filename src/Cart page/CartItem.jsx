import PropTypes from "prop-types";
import { useState } from "react";
import { useOutletContext } from "react-router-dom";

const CartItem = ({ cartItem, cartItemIndex }) => {
  const { cart, setCart } = useOutletContext();
  const {
    productDetails: { id, title, image, category, price },
    quantity,
  } = cartItem;

  const [productQuantity, setProductQuantity] = useState(quantity);

  const totalItemPrice = price * productQuantity;

  const handleIncreaseQuantity = () => {
    const newCart = structuredClone(cart);

    if (!productQuantity) {
      newCart[cartItemIndex].quantity = 1;
      setProductQuantity(1);
      setCart(newCart);
      return;
    }

    newCart[cartItemIndex].quantity += 1;
    setProductQuantity((value) => Number.parseInt(value) + 1);
    setCart(newCart);
  };

  const handleDecreaseQuantity = () => {
    const newCart = structuredClone(cart);

    if (productQuantity <= 1 || !productQuantity) {
      newCart[cartItemIndex].quantity = 1;
      setProductQuantity(1);
      setCart(newCart);
      return;
    }

    newCart[cartItemIndex].quantity -= 1;
    setProductQuantity((value) => Number.parseInt(value) - 1);
    setCart(newCart);
  };

  const handleTypeQuantity = (e) => {
    const typedValue = e.target.value;
    const convertedNumber = Number.parseInt(typedValue);
    const newQuantity =
      !isNaN(convertedNumber) && convertedNumber >= 1 ? convertedNumber : 0;

    const newCart = structuredClone(cart);
    newCart[cartItemIndex].quantity = newQuantity;

    setCart(newCart);
    setProductQuantity(typedValue);
  };

  const handleRemoveCartItem = () => {
    const newCart = cart.filter(
      (cartItem) => cartItem.productDetails.id !== id
    );
    setCart(newCart);
  };

  return (
    <li>
      <div>
        <div>
          <img src={image} alt="" />
          <div>
            <p>{title}</p>
            <p>Category: {category}</p>
          </div>
        </div>
        <div>
          <button
            aria-label="Increase quantity"
            onClick={handleIncreaseQuantity}
          >
            +
          </button>
          <input
            aria-label="Quantity value"
            type="number"
            min={1}
            value={productQuantity}
            onChange={handleTypeQuantity}
          />
          <button
            aria-label="Decrease quantity"
            onClick={handleDecreaseQuantity}
          >
            -
          </button>
        </div>
        <div>${price}</div>
        <div>${totalItemPrice}</div>
      </div>
      <button aria-label="Remove item from cart" onClick={handleRemoveCartItem}>
        X
      </button>
    </li>
  );
};

CartItem.propTypes = {
  cartItem: PropTypes.shape({
    productDetails: PropTypes.shape({
      id: PropTypes.number,
      title: PropTypes.string,
      image: PropTypes.string,
      category: PropTypes.string,
      price: PropTypes.number,
    }),
    quantity: PropTypes.number,
  }),
  cartItemIndex: PropTypes.number.isRequired,
};

export default CartItem;
