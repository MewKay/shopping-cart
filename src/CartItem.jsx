import PropTypes from "prop-types";

const CartItem = ({ cartItem }) => {
  const {
    productDetails: { title, image, category, price },
    quantity,
  } = cartItem;
  const totalItemPrice = price * quantity;

  return (
    <li>
      <div>
        <div>
          <img src={image} alt="" />
          <div>
            <p>{title}</p>
            <p>Category: {category}</p>
          </div>
        </div>
        <div>
          <button>+</button>
          <input value={quantity} />
          <button>-</button>
        </div>
        <div>${price}</div>
        <div>${totalItemPrice}</div>
      </div>
      <button>X</button>
    </li>
  );
};

CartItem.propTypes = {
  cartItem: PropTypes.shape({
    productDetails: PropTypes.shape({
      title: PropTypes.string,
      image: PropTypes.string,
      category: PropTypes.string,
      price: PropTypes.number,
    }),
    quantity: PropTypes.number,
  }),
};

export default CartItem;
