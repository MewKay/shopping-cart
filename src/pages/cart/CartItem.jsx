import PropTypes from "prop-types";
import { useState } from "react";
import { useOutletContext } from "react-router-dom";
import { X } from "lucide-react";
import styles from "./CartItem.module.css";
import QuantitySelector from "../../components/quantity-selector/QuantitySelector";

const CartItem = ({ cartItem, cartItemIndex }) => {
  const { cart, setCart } = useOutletContext();
  const {
    productDetails: { id, title, image, category, price },
    quantity,
  } = cartItem;

  const [productQuantity, setProductQuantity] = useState(quantity);

  //Will format the number to always have 2 decimal places
  const totalItemPrice = Number(
    Math.round(price * productQuantity + "e2") + "e-2"
  );

  const handleIncreaseQuantity = () => {
    const newCart = structuredClone(cart);

    if (!productQuantity) {
      newCart[cartItemIndex].quantity = 1;
      setProductQuantity(1);
      setCart(newCart);
      return;
    }

    newCart[cartItemIndex].quantity += 1;
    setProductQuantity((value) => Number.parseInt(value) + 1);
    setCart(newCart);
  };

  const handleDecreaseQuantity = () => {
    const newCart = structuredClone(cart);

    if (productQuantity <= 1 || !productQuantity) {
      newCart[cartItemIndex].quantity = 1;
      setProductQuantity(1);
      setCart(newCart);
      return;
    }

    newCart[cartItemIndex].quantity -= 1;
    setProductQuantity((value) => Number.parseInt(value) - 1);
    setCart(newCart);
  };

  const handleTypeQuantity = (e) => {
    const typedValue = e.target.value;
    const convertedNumber = Number.parseInt(typedValue);
    const newQuantity =
      !isNaN(convertedNumber) && convertedNumber >= 1 ? convertedNumber : 0;

    const newCart = structuredClone(cart);
    newCart[cartItemIndex].quantity = newQuantity;

    setCart(newCart);
    setProductQuantity(typedValue);
  };

  const handleRemoveCartItem = () => {
    const newCart = cart.filter(
      (cartItem) => cartItem.productDetails.id !== id
    );
    setCart(newCart);
  };

  return (
    <>
      <div className={styles["details-container"]}>
        <div className={styles["image-container"]}>
          <img src={image} alt="" className={styles["image-product"]} />
        </div>
        <div className={styles["details-info"]}>
          <p className={styles["details-title"]}>{title}</p>
          <p className={styles["details-category"]}>Category: {category}</p>
        </div>
      </div>
      <QuantitySelector
        productQuantity={productQuantity}
        onIncreaseQuantity={handleIncreaseQuantity}
        onTypeQuantity={handleTypeQuantity}
        onDecreaseQuantity={handleDecreaseQuantity}
      />
      <div>${price}</div>
      <div>${totalItemPrice}</div>
      <button
        aria-label="Remove item from cart"
        onClick={handleRemoveCartItem}
        className={styles["remove-item"]}
      >
        <X />
      </button>
    </>
  );
};

CartItem.propTypes = {
  cartItem: PropTypes.shape({
    productDetails: PropTypes.shape({
      id: PropTypes.number,
      title: PropTypes.string,
      image: PropTypes.string,
      category: PropTypes.string,
      price: PropTypes.number,
    }),
    quantity: PropTypes.number,
  }),
  cartItemIndex: PropTypes.number.isRequired,
};

export default CartItem;
