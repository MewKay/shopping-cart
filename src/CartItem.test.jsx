import { describe, expect, it } from "vitest";
import { render } from "@testing-library/react";
import CartItem from "./CartItem";

//Component props
const cartItem = {
  productDetails: {
    id: 0,
    title: "Product 0",
    image: "https://image.placeholder.com/product",
    category: "jewelery",
    price: 50,
  },
  quantity: 4,
};

describe("CartItem component", () => {
  it("renders correctly", () => {
    const { container } = render(<CartItem cartItem={cartItem} />);

    expect(container).toMatchSnapshot();
  });
});
