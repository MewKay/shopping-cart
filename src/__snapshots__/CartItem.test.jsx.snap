// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CartItem component > renders correctly 1`] = `
<div>
  <li>
    <div>
      <div>
        <img
          alt=""
          src="https://image.placeholder.com/product"
        />
        <div>
          <p>
            Product 0
          </p>
          <p>
            Category: 
            jewelery
          </p>
        </div>
      </div>
      <div>
        <button>
          +
        </button>
        <input
          value="4"
        />
        <button>
          -
        </button>
      </div>
      <div>
        $
        50
      </div>
      <div>
        $
        200
      </div>
    </div>
    <button>
      X
    </button>
  </li>
</div>
`;
